src/esssans/normalization.py

```diff
@@ -64,6 +65,68 @@ def solid_angle_rectangular_approximation(
     )
 
 
+def solid_angle(
+    data: CalibratedMaskedData[RunType],
+    detector: DetectorGeometry[RunType],
+) -> SolidAngle[RunType]:
+    """
+    Solid angle for cylindrical pixels.
+
+    Note that the approximation is valid when the distance from sample
+    to pixel is much larger than the length or the radius of the pixels.
+
+    Parameters
+    ----------
+    data:
+        The DataArray that contains the positions of the detector pixels in coords.
+        The positions must be in the sample reference frame.
+
+    detector:
+        The DataGroup that contains the geometry of the detector.
+        Must contain pixel_shape and the transformation from the
+        local reference frame of the detector to the sample reference frame.
+
+    Returns
+    -------
+    :
+        The solid angle of the detector pixels, as viewed from the sample position.
+    """
+    face_1_center, face_1_edge, face_2_center = (
+        detector['pixel_shape']['vertices']['vertex', i] for i in range(3)
+    )
+    cylinder_axis = (
+        detector['transform'] * face_2_center - detector['transform'] * face_1_center
+    )
+    radius = sc.norm(face_1_center - face_1_edge)
+    length = sc.norm(cylinder_axis)
+
+    return SolidAngle[RunType](
+        approximate_solid_angle_for_cylinder_shaped_pixel_of_detector(
+            pixel_position=data.coords['position'],
+            cylinder_axis=cylinder_axis,
+            radius=radius,
+            length=length,
+        )
+    )
+
+
+def approximate_solid_angle_for_cylinder_shaped_pixel_of_detector(
+    pixel_position: sc.Variable,
+    cylinder_axis: sc.Variable,
+    radius: sc.Variable,
+    length: sc.Variable,
+):
+    """Computes solid angle of a detector pixel under the assumption that the
+    distance to the detector pixel from the sample is large compared to the radius
+    and the length of the piece of the detector cylinder that makes up the pixel.
+    """
+    norm_pp = sc.norm(pixel_position)
+    norm_ca = sc.norm(cylinder_axis)
+    cosalpha = sc.norm(sc.cross(pixel_position, cylinder_axis)) / (norm_pp * norm_ca)
```

Comment:
$\alpha$ is the angle between the surface normal and `pixel_position`.
The angle between `pixel_position` and `cylinder_axis` is $\theta = \pi/2 - \alpha$.

$ \sin(\theta) = \sin(\pi/2 - \alpha) = \cos(\alpha)$, if $\theta, \alpha \le \pi/2$

Reply:
Can you put that in a comment?

Reply:
The definition of solid angle is $\Omega = \int_S \frac{\hat{r} \cdot n}{r^2}\ dS$ where $n$ is the normal of the surface and in our case $S$ is a cylinder.

The cylinder surface is approximated by a rectangle cross section of the cylinder, called $\hat{S}$.

The solid angle of the rectangle $\int_\hat{S} \frac{\hat{r} \cdot n}{r^2}\ dS $ is approximated as $ A_{\hat{S}} \frac{\hat{r}_{midpoint}\cdot n_{midpoint}}{r_{midpoint}}$ under the assumption that $r$ changes little over the integral.

Reply:
Which surface normal?

Reply:
In the approximation, the surface normal is the surface normal of the rectangle.
In the equation for the exact solid angle, the surface normal is the surface normal of the cylinder.

Reply:
In any case, yes, that makes sense, I think.

Would it make any performance difference if we used the dot-product instead to compute the same? It would avoid creating an intermediate array of vectors (and have an array of scalars instead)?

Reply:
**How is the rectangle approximating the cylinder surface selected?**

It is selected as the rectangle contained in $S$, parallel to the cylinder axis, with maximal solid angle. This makes $\Omega(\hat{S}) \approx \Omega(S)$ when $||r||$ is large compared to $R$ (the radius of the cylinder). In the below, let $c$ be the cylinder axis, $n$ the rectangle normal, and $\hat{r}$ the unit vector to the center of the cylinder.

1. $\hat{S}$ is parallel to the axis of the cylinder: $n \cdot c = 0$
2. $\hat{S}$ is selected to maximize $\hat{r} \cdot n$. Under the constraint in (1) this occurs when $n$, $\hat{r}$ and $c$ all lie in the same plane.

Call the angle between $n$ and $\hat{r}$ $\alpha$, the angle between $c$ and $\hat{r}$ $\theta$.

* From (1): The angle between $n$ and $c$ is $\pi/2$.
* From (2): $\alpha + \theta + \pi/2 = \pi \ \iff \theta = \pi/2 - \alpha$  

$\hat{r} \times \hat{c} = \sin(\theta) = \sin(\pi/2 - \alpha)$ $=\cos(\alpha) = \hat{r} \cdot \hat{n}  $

Reply:
> Would it make any performance difference if we used the dot-product instead to compute the same? It would avoid creating an intermediate array of vectors (and have an array of scalars instead)?

Yes maybe there's a better way.
We could do:

From 1 and 2: $r = (r \cdot n) n + (r \cdot c) c$

$r \cdot r = (r \cdot n)^2 +  (r \cdot c)^2 $
$|r \cdot n| = \sqrt{1 -  (r \cdot c)^2} $

This looks like what you talked about before.

Reply:
Out of interest, did you check which version is implemented in Mantid?

Can you add the math and explanation as a docstring?

Reply:
> Out of interest, did you check which version is implemented in Mantid?

I didn't understand what they did there before. But now I think I see that they do.
First, they work with the non-normalized position vector $r$ instead of $\hat{r}$.
They also project $r$ on the plane orthogonal to $c$:
$p := r - (r \cdot c)c = (r\cdot n) n$

But then they do slightly different, they return
$\frac{p \cdot r}{||r|| \ ||p||} = \frac{(r \cdot n)^2}{||r||^2cos(\alpha)} = cos(\alpha)$

In terms of efficiency I think this version and the one we have now are about the same.
It looks here like they have two scalar products but they don't because the projection on the plane orthogonal to $c$ is just setting the y-component (or x-component) to 0 in their implementation. They can do that because it is assumed that the detector axis is either horizontal or vertical.

Reply:
We can ignore `scaleFactor`: This comes from the IDF description of the pixel shape, which may define a unit-cylinder, which is then scaled to the actual size. In NeXus there is not such scale, the actual size is stored directly.